internal/server/logger.go

```diff
@@ -0,0 +1,127 @@
+}
+
+func NewStructuredLogger(handler slog.Handler, options *StructuredLoggerOptions) func(next http.Handler) http.Handler {
+	if options == nil {
+		options = DefaultStructuredLoggerOptions
+	}
+
+	return middleware.RequestLogger(&StructuredLogger{Logger: handler, Options: options})
+}
+
+func (l *StructuredLogger) NewLogEntry(r *http.Request) middleware.LogEntry {
+	logFields := []slog.Attr{
+		slog.String(l.Options.TimeFieldName, time.Now().UTC().Format(l.Options.TimeFieldFormat)),
+	}
+
+	if reqID := middleware.GetReqID(r.Context()); reqID != "" {
+		logFields = append(logFields, slog.String("req_id", reqID))
+	}
+
+	handler := l.Logger.WithAttrs(append(logFields,
+		slog.String("http_scheme", r.URL.Scheme),
+		slog.String("http_proto", r.Proto),
+		slog.String("http_method", r.Method),
+		slog.String("remote_addr", r.RemoteAddr),
+		slog.String("user_agent", r.UserAgent()),
+		slog.String("uri", r.URL.String())))
+
+	return &StructuredLoggerEntry{logger: slog.New(handler), request: r, level: l.Options.LogLevel}
+}
+
+type StructuredLoggerEntry struct {
+	// logger is the underlying logger.
+	logger *slog.Logger
+
+	// request is the original request, stored for later use when writing the log entry.
+	request *http.Request
+
+	// level is the default log level of the entry.
+	level slog.Level
+}
+
+func (e *StructuredLoggerEntry) statusLevel(status int) slog.Level {
+	switch {
+	case status < 400:
+		return e.level
+	case status >= 400 && status < 500 && status != http.StatusNotFound:
+		return slog.LevelWarn
+	case status >= 500:
+		return slog.LevelError
+	default:
+		return e.level
+	}
+}
+
+func (e *StructuredLoggerEntry) Write(status, bytes int, header http.Header, elapsed time.Duration, extra interface{}) {
+	e.logger.LogAttrs(e.request.Context(), e.statusLevel(status), "request complete",
+		slog.Int("resp_status", status),
+		slog.Int("resp_byte_length", bytes),
+		slog.Float64("resp_elapsed_ms", float64(elapsed.Nanoseconds())/1000000.0),
+	)
+}
+
+func (e *StructuredLoggerEntry) Panic(v interface{}, stack []byte) {
+	e.logger.LogAttrs(e.request.Context(), e.statusLevel(500), "request panicked",
+		slog.String("stack", string(stack)),
+		slog.String("panic", fmt.Sprintf("%+v", v)),
+	)
+}
+
+// Helper methods used by the application to get the request-scoped
+// logger entry and set additional fields between handlers.
+
+func GetLogEntry(r *http.Request) *slog.Logger {
+	entry := middleware.GetLogEntry(r).(*StructuredLoggerEntry)
+	return entry.logger
+}
+
+func LogEntrySetField(r *http.Request, key string, value interface{}) {
+	if entry, ok := r.Context().Value(middleware.LogEntryCtxKey).(*StructuredLoggerEntry); ok {
+		entry.logger = entry.logger.With(key, value)
+	}
+}
+
+func LogEntrySetFields(r *http.Request, fields map[string]interface{}) {
+	if entry, ok := r.Context().Value(middleware.LogEntryCtxKey).(*StructuredLoggerEntry); ok {
+		for k, v := range fields {
+			entry.logger = entry.logger.With(k, v)
+		}
+	}
+}
```

Comment:
I know it's fairly contested in the community, but a majority of trustable sources suggests this practice is as bad as it gets. It abuses the context as a request scoped service locator.

The primary reason why the log handler receives the context is that it can unpack potential details necessary for logs _from_ the context.

The best practice is to inject the logger (slog.Logger) wherever necessary and use a handler that's capable of extracting relevant information from the context.